File: unidbg-android/src/main/java/com/vortexdbg/linux/android/ElfLibraryFile.java
package com.vortexdbg.linux.android;

import com.vortexdbg.Emulator;
import com.vortexdbg.Utils;
import com.vortexdbg.spi.LibraryFile;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;

public class ElfLibraryFile implements LibraryFile {

    private final File elfFile;
    private final boolean is64Bit;

    public ElfLibraryFile(File elfFile, boolean is64Bit) {
        this.elfFile = elfFile;
        this.is64Bit = is64Bit;
    }

    @Override
    public long getFileSize() {
        return elfFile.length();
    }

    @Override
    public String getName() {
        return elfFile.getName();
    }

    @Override
    public String getMapRegionName() {
        return getPath();
    }

    @Override
    public LibraryFile resolveLibrary(Emulator<?> emulator, String soName) {
        File file = new File(elfFile.getParentFile(), soName);
        return file.canRead() ? new ElfLibraryFile(file, is64Bit) : null;
    }

    @Override
    public ByteBuffer mapBuffer() throws IOException {
        return Utils.mapBuffer(elfFile);
    }

    @Override
    public String getPath() {
        String name = getName();
        if (name.endsWith(".so")) {
            return "/system/" + (is64Bit ? "lib64/" : "lib/") + name;
        } else {
            return "/system/bin/" + name;
        }
    }

}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/android/ElfLibraryFile.kt
package com.vortexdbg.linux.android

import com.vortexdbg.Emulator
import com.vortexdbg.Utils
import com.vortexdbg.spi.LibraryFile

import java.io.File
import java.io.IOException
import java.nio.ByteBuffer

open class ElfLibraryFile(private val elfFile: File, private val is64Bit: Boolean) : LibraryFile {

    override fun getFileSize(): Long {
        return elfFile.length()
    }

    override fun getName(): String {
        return elfFile.name
    }

    override fun getMapRegionName(): String {
        return getPath()
    }

    override fun resolveLibrary(emulator: Emulator<*>, soName: String): LibraryFile? {
        val file = File(elfFile.parentFile, soName)
        return if (file.canRead()) ElfLibraryFile(file, is64Bit) else null
    }

    @Throws(IOException::class)
    override fun mapBuffer(): ByteBuffer {
        return Utils.mapBuffer(elfFile)
    }

    override fun getPath(): String {
        val name = getName()
        return if (name.endsWith(".so")) {
            "/system/" + (if (is64Bit) "lib64/" else "lib/") + name
        } else {
            "/system/bin/$name"
        }
    }

}
